File: src/osgui/osx/ospopup.m
/*
 * NAppGUI Cross-platform C SDK
 * 2015-2025 Francisco Garcia Collado
 * MIT Licence
 * https://nappgui.com/en/legal/license.html
 *
 * File: ospopup.m
 *
 */

/* Operating System native popup button */

#include "ospopup_osx.inl"
#include "oscontrol_osx.inl"
#include "ospanel_osx.inl"
#include "oswindow_osx.inl"
#include "../ospopup.h"
#include "../osgui.inl"
#include <draw2d/image.h>
#include <core/event.h>
#include <core/heap.h>
#include <sewer/cassert.h>
#include <sewer/ptr.h>

#if !defined(__MACOS__)
#error This file is only for OSX
#endif

/*---------------------------------------------------------------------------*/

@interface OSXPopUp : NSPopUpButton
{
  @public
    OSTextAttr attrs;
    Listener *OnSelect;
}
@end

/*---------------------------------------------------------------------------*/

@implementation OSXPopUp

/*---------------------------------------------------------------------------*/

- (IBAction)onSelectionChange:(id)sender
{
    cassert(sender == self);
    if ([self isEnabled] == YES && self->OnSelect != NULL)
    {
        EvButton params;
        params.state = ekGUI_ON;
        params.index = (uint32_t)[self indexOfSelectedItem];
        params.text = NULL; /*(const char_t*)[[self titleOfSelectedItem] UTF8String];*/
        listener_event(self->OnSelect, ekGUI_EVENT_POPUP, cast(sender, OSPopUp), &params, NULL, OSPopUp, EvButton, void);
    }
}

/*---------------------------------------------------------------------------*/

- (void)mouseDown:(NSEvent *)theEvent
{
    if (_oswindow_mouse_down(cast(self, OSControl)) == TRUE)
        [super mouseDown:theEvent];
}

@end

/*---------------------------------------------------------------------------*/

OSPopUp *ospopup_create(const uint32_t flags)
{
    OSXPopUp *popup = nil;
    NSPopUpButtonCell *cell = nil;
    unref(flags);
    heap_auditor_add("OSXPopUp");
    popup = [[OSXPopUp alloc] initWithFrame:NSZeroRect];
    popup->OnSelect = NULL;
    _oscontrol_init(popup);
    _oscontrol_init_textattr(&popup->attrs);
    _oscontrol_set_font(popup, &popup->attrs, popup->attrs.font);
    cell = [popup cell];
    /*[cell setBezelStyle:NSShadowlessSquareBezelStyle];*/
    [cell setArrowPosition:NSPopUpArrowAtBottom];
    _oscontrol_cell_set_control_size(cell, ekGUI_SIZE_REGULAR);
    [popup setPullsDown:NO];
    [popup setTarget:popup];
    [popup setAction:@selector(onSelectionChange:)];
    return cast(popup, OSPopUp);
}

/*---------------------------------------------------------------------------*/

void ospopup_destroy(OSPopUp **popup)
{
    OSXPopUp *lpopup = nil;
    cassert_no_null(popup);
    lpopup = *dcast(popup, OSXPopUp);
    cassert_no_null(lpopup);
    listener_destroy(&lpopup->OnSelect);
    _oscontrol_remove_textattr(&lpopup->attrs);
    [lpopup release];
    *popup = NULL;
    heap_auditor_delete("OSXPopUp");
}

/*---------------------------------------------------------------------------*/

void ospopup_OnSelect(OSPopUp *popup, Listener *listener)
{
    cassert_no_null(popup);
    listener_update(&cast(popup, OSXPopUp)->OnSelect, listener);
}

/*---------------------------------------------------------------------------*/

static void i_add_elem(OSPopUp *popup, const char_t *text, const Image *image)
{
    OSXPopUp *lpopup = cast(popup, OSXPopUp);
    NSString *str = nil;
    NSMenuItem *item = nil;
    cassert_no_null(lpopup);
    cassert_no_null(text);
    str = [[NSString alloc] initWithUTF8String:cast_const(text, char)];
    item = [[NSMenuItem alloc] initWithTitle:str action:nil keyEquivalent:@""];
    [item setTarget:lpopup];
    [[lpopup menu] addItem:item];
    [str release];
    [item release];

    if (image != NULL)
    {
        NSInteger num_items;
        NSMenuItem *item = nil;
        NSImage *nsimage = cast(image_native(image), NSImage);
#if defined(__ASSERTS__)
        NSSize size = [nsimage size];
        cassert(size.width == 16.f && size.height == 16.f);
#endif

        num_items = [lpopup numberOfItems];
        cassert(num_items > 0);
        item = [lpopup itemAtIndex:num_items - 1];
        [item setImage:nsimage];
        [item setOnStateImage:nil];
        [item setMixedStateImage:nil];
    }
}

/*---------------------------------------------------------------------------*/

void ospopup_elem(OSPopUp *popup, const ctrl_op_t op, const uint32_t idx, const char_t *text, const Image *image)
{
    if (op == ekCTRL_OP_ADD)
    {
        i_add_elem(popup, text, image);
    }
    else if (op == ekCTRL_OP_SET)
    {
        NSArray *items = nil;
        NSMenuItem *item = nil;
        cassert_no_null(popup);
        items = [cast(popup, OSXPopUp) itemArray];
        cassert_no_null(items);
        cassert(idx < [items count]);
        item = [items objectAtIndex:(NSUInteger)idx];

        {
            NSString *str = [[NSString alloc] initWithUTF8String:cast_const(text, char)];
            [item setTitle:str];
            [str release];
        }

#if defined(__ASSERTS__)
        if (image != NULL)
        {
            NSSize size = [cast(image_native(image), NSImage) size];
            cassert(size.width == 16.f && size.height == 16.f);
        }
#endif

        [item setImage:image != NULL ? cast(image_native(image), NSImage) : nil];
    }
    else if (op == ekCTRL_OP_DEL)
    {
        [cast(popup, OSXPopUp) removeItemAtIndex:(NSInteger)idx];
    }
    else
    {
        cassert_msg(FALSE, "Not implemented");
    }
}

/*---------------------------------------------------------------------------*/

void ospopup_tooltip(OSPopUp *popup, const char_t *text)
{
    unref(popup);
    unref(text);
}

/*---------------------------------------------------------------------------*/

void ospopup_font(OSPopUp *popup, const Font *font)
{
    OSXPopUp *lpopup = cast(popup, OSXPopUp);
    cassert_no_null(lpopup);
    _oscontrol_set_font(lpopup, &lpopup->attrs, font);
}

/*---------------------------------------------------------------------------*/

void ospopup_list_height(OSPopUp *popup, const uint32_t num_elems)
{
    unref(popup);
    unref(num_elems);
}

/*---------------------------------------------------------------------------*/

void ospopup_selected(OSPopUp *popup, const uint32_t lindex)
{
    OSXPopUp *lpopup = cast(popup, OSXPopUp);
    cassert_no_null(lpopup);
    if (lindex != UINT32_MAX)
    {
        cassert((NSInteger)lindex <= [lpopup numberOfItems]);
        [lpopup selectItemAtIndex:(NSInteger)lindex];
    }
    else
    {
        [lpopup selectItemAtIndex:-1];
    }
}

/*---------------------------------------------------------------------------*/

uint32_t ospopup_get_selected(const OSPopUp *popup)
{
    OSXPopUp *lpopup = cast(popup, OSXPopUp);
    cassert_no_null(lpopup);
    return (uint32_t)[lpopup indexOfSelectedItem];
}

/*---------------------------------------------------------------------------*/

void ospopup_bounds(const OSPopUp *popup, const char_t *text, real32_t *width, real32_t *height)
{
    OSXPopUp *lpopup = cast(popup, OSXPopUp);
    cassert_no_null(lpopup);
    cassert_no_null(width);
    cassert_no_null(height);
    font_extents(lpopup->attrs.font, text, -1.f, width, height);
    *width += 40.f;
    *height = 24.f;
}

/*---------------------------------------------------------------------------*/

void ospopup_attach(OSPopUp *popup, OSPanel *panel)
{
    _ospanel_attach_control(panel, cast(popup, NSView));
}

/*---------------------------------------------------------------------------*/

void ospopup_detach(OSPopUp *popup, OSPanel *panel)
{
    _ospanel_detach_control(panel, cast(popup, NSView));
}

/*---------------------------------------------------------------------------*/

void ospopup_visible(OSPopUp *popup, const bool_t visible)
{
    _oscontrol_set_visible(cast(popup, NSView), visible);
}

/*---------------------------------------------------------------------------*/

void ospopup_enabled(OSPopUp *popup, const bool_t enabled)
{
    _oscontrol_set_enabled(cast(popup, NSControl), enabled);
}

/*---------------------------------------------------------------------------*/

void ospopup_size(const OSPopUp *popup, real32_t *width, real32_t *height)
{
    _oscontrol_get_size(cast(popup, NSView), width, height);
}

/*---------------------------------------------------------------------------*/

void ospopup_origin(const OSPopUp *popup, real32_t *x, real32_t *y)
{
    _oscontrol_get_origin(cast(popup, NSView), x, y);
}

/*---------------------------------------------------------------------------*/

void ospopup_frame(OSPopUp *popup, const real32_t x, const real32_t y, const real32_t width, const real32_t height)
{
    _oscontrol_set_frame(cast(popup, NSView), x, y, width, height);
}

/*---------------------------------------------------------------------------*/

BOOL _ospopup_is(NSView *view)
{
    return [view isKindOfClass:[OSXPopUp class]];
}

File: src/osgui/win/ospopup.c
/*
 * NAppGUI Cross-platform C SDK
 * 2015-2025 Francisco Garcia Collado
 * MIT Licence
 * https://nappgui.com/en/legal/license.html
 *
 * File: ospopup.c
 *
 */

/* Operating System native popup button */

#include "osgui_win.inl"
#include "oscontrol_win.inl"
#include "oscombo_win.inl"
#include "ospanel_win.inl"
#include "oswindow_win.inl"
#include "osimglist.inl"
#include "ostooltip.inl"
#include "../ospopup.h"
#include "../ospopup.inl"
#include "../osgui.inl"
#include <draw2d/font.h>
#include <core/event.h>
#include <core/heap.h>
#include <sewer/cassert.h>

#if !defined(__WINDOWS__)
#error This file is only for Windows
#endif

struct _ospopup_t
{
    OSControl control;
    Font *font;
    HWND combo_hwnd;
    WNDPROC def_combo_proc;
    uint32_t list_num_elems;
    OSImgList *image_list;
    Listener *OnSelect;
};

/*---------------------------------------------------------------------------*/

static LRESULT CALLBACK i_WndProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
{
    OSPopUp *popup = cast(GetWindowLongPtr(hwnd, GWLP_USERDATA), OSPopUp);
    cassert_no_null(popup);

    switch (uMsg)
    {
    case WM_ERASEBKGND:
        return 1;

    case WM_LBUTTONDOWN:
    case WM_LBUTTONDBLCLK:
        if (_oswindow_mouse_down(cast(popup, OSControl)) == TRUE)
            break;
        return 0;
    }

    return CallWindowProc(popup->control.def_wnd_proc, hwnd, uMsg, wParam, lParam);
}

/*---------------------------------------------------------------------------*/

static LRESULT CALLBACK i_ComboWndProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
{
    OSPopUp *popup = cast(GetWindowLongPtr(hwnd, GWLP_USERDATA), OSPopUp);
    cassert_no_null(popup);

    switch (uMsg)
    {
    case WM_LBUTTONDOWN:
    case WM_LBUTTONDBLCLK:
        if (_oswindow_mouse_down(cast(popup, OSControl)) == TRUE)
            break;
        return 0;
    }

    return popup->def_combo_proc(hwnd, uMsg, wParam, lParam);
}

/*---------------------------------------------------------------------------*/

OSPopUp *ospopup_create(const uint32_t flags)
{
    OSPopUp *popup = NULL;
    DWORD dwStyle = WS_CHILD | WS_CLIPSIBLINGS | CBS_DROPDOWNLIST;
    unref(flags);
    popup = heap_new0(OSPopUp);
    popup->control.type = ekGUI_TYPE_POPUP;
    _oscontrol_init(cast(popup, OSControl), PARAM(dwExStyle, WS_EX_NOPARENTNOTIFY | CBES_EX_NOSIZELIMIT), dwStyle, WC_COMBOBOXEX, 0, 120, i_WndProc, kDEFAULT_PARENT_WINDOW);
    popup->font = _osgui_create_default_font();
    popup->combo_hwnd = (HWND)SendMessage(popup->control.hwnd, CBEM_GETCOMBOCONTROL, (WPARAM)0, (LPARAM)0);
    popup->def_combo_proc = (WNDPROC)SetWindowLongPtr(popup->combo_hwnd, GWLP_WNDPROC, (LONG_PTR)i_ComboWndProc);
    SetWindowLongPtr(popup->combo_hwnd, GWLP_USERDATA, (LONG_PTR)popup);
    popup->image_list = _osimglist_create(16);
    popup->list_num_elems = 5;
    _oscontrol_set_font(cast(popup, OSControl), popup->font);
    return popup;
}

/*---------------------------------------------------------------------------*/

void ospopup_destroy(OSPopUp **popup)
{
    cassert_no_null(popup);
    cassert_no_null(*popup);
    font_destroy(&(*popup)->font);
    listener_destroy(&(*popup)->OnSelect);
    _osimglist_destroy(&(*popup)->image_list);
    _oscontrol_destroy(&(*popup)->control);
    heap_delete(popup, OSPopUp);
}

/*---------------------------------------------------------------------------*/

void ospopup_OnSelect(OSPopUp *popup, Listener *listener)
{
    cassert_no_null(popup);
    listener_update(&popup->OnSelect, listener);
}

/*---------------------------------------------------------------------------*/

void ospopup_elem(OSPopUp *popup, const ctrl_op_t op, const uint32_t index, const char_t *text, const Image *image)
{
    cassert_no_null(popup);
    _oscombo_elem(popup->control.hwnd, popup->image_list, op, index, text, image);
    if (SendMessage(popup->control.hwnd, CB_GETCURSEL, (WPARAM)0, (LPARAM)0) == -1)
        SendMessage(popup->control.hwnd, CB_SETCURSEL, (WPARAM)0, (LPARAM)0);
    InvalidateRect(popup->control.hwnd, NULL, FALSE);
}

/*---------------------------------------------------------------------------*/

void ospopup_tooltip(OSPopUp *popup, const char_t *text)
{
    cassert_no_null(popup);
    _ostooltip_set_text(&popup->control.tooltip_hwnd, popup->combo_hwnd, text);
}

/*---------------------------------------------------------------------------*/

void ospopup_font(OSPopUp *popup, const Font *font)
{
    cassert_no_null(popup);
    _oscontrol_update_font(cast(popup, OSControl), &popup->font, font);
}

/*---------------------------------------------------------------------------*/

void ospopup_list_height(OSPopUp *popup, const uint32_t num_elems)
{
    cassert_no_null(popup);
    popup->list_num_elems = num_elems;
    _oscombo_set_list_height(popup->control.hwnd, popup->combo_hwnd, _osimglist_height(popup->image_list), popup->list_num_elems);
}

/*---------------------------------------------------------------------------*/

void ospopup_selected(OSPopUp *popup, const uint32_t index)
{
    cassert_no_null(popup);
    if (index != UINT32_MAX)
    {
        LRESULT ret = SendMessage(popup->control.hwnd, CB_SETCURSEL, (WPARAM)index, (LPARAM)0);
        cassert_unref(ret == (LRESULT)index, ret);
    }
    else
    {
        LRESULT ret = SendMessage(popup->control.hwnd, CB_SETCURSEL, (WPARAM)-1, (LPARAM)0);
        cassert_unref(ret == (LRESULT)-1, ret);
    }
}

/*---------------------------------------------------------------------------*/

uint32_t ospopup_get_selected(const OSPopUp *popup)
{
    cassert_no_null(popup);
    return (uint32_t)SendMessage(popup->control.hwnd, CB_GETCURSEL, (WPARAM)0, (LPARAM)0);
}

/*---------------------------------------------------------------------------*/

void ospopup_bounds(const OSPopUp *popup, const char_t *text, real32_t *width, real32_t *height)
{
    uint32_t imgwidth = UINT32_MAX;
    cassert_no_null(popup);
    cassert_no_null(width);
    cassert_no_null(height);
    font_extents(popup->font, text, -1.f, width, height);
    *width += 40.f;
    *height += 8.f;
    imgwidth = _osimglist_width(popup->image_list);
    if (imgwidth != UINT32_MAX)
        *width += (real32_t)(imgwidth + 5);
}

/*---------------------------------------------------------------------------*/

void ospopup_attach(OSPopUp *popup, OSPanel *panel)
{
    _ospanel_attach_control(panel, cast(popup, OSControl));
}

/*---------------------------------------------------------------------------*/

void ospopup_detach(OSPopUp *popup, OSPanel *panel)
{
    _ospanel_detach_control(panel, cast(popup, OSControl));
}

/*---------------------------------------------------------------------------*/

void ospopup_visible(OSPopUp *popup, const bool_t visible)
{
    _oscontrol_set_visible(cast(popup, OSControl), visible);
}

/*---------------------------------------------------------------------------*/

void ospopup_enabled(OSPopUp *popup, const bool_t enabled)
{
    _oscontrol_set_enabled(cast(popup, OSControl), enabled);
}

/*---------------------------------------------------------------------------*/

void ospopup_size(const OSPopUp *popup, real32_t *width, real32_t *height)
{
    _oscontrol_get_size(cast_const(popup, OSControl), width, height);
}

/*---------------------------------------------------------------------------*/

void ospopup_origin(const OSPopUp *popup, real32_t *x, real32_t *y)
{
    _oscontrol_get_origin(cast_const(popup, OSControl), x, y);
}

/*---------------------------------------------------------------------------*/

void ospopup_frame(OSPopUp *popup, const real32_t x, const real32_t y, const real32_t width, const real32_t height)
{
    _oscontrol_set_frame(cast(popup, OSControl), x, y, width, height);
    _oscombo_set_list_height(popup->control.hwnd, popup->combo_hwnd, _osimglist_height(popup->image_list), popup->list_num_elems);
}

/*---------------------------------------------------------------------------*/

void _ospopup_command(OSPopUp *popup, WPARAM wParam)
{
    cassert_no_null(popup);
    if (HIWORD(wParam) == CBN_SELCHANGE && IsWindowEnabled(popup->control.hwnd) && popup->OnSelect != NULL)
    {
        EvButton params;
        params.state = ekGUI_ON;
        params.index = (uint32_t)SendMessage(popup->control.hwnd, CB_GETCURSEL, (WPARAM)0, (LPARAM)0);
        cassert(params.index < (uint32_t)SendMessage(popup->control.hwnd, CB_GETCOUNT, (WPARAM)0, (LPARAM)0));
        params.text = NULL;
        listener_event(popup->OnSelect, ekGUI_EVENT_POPUP, popup, &params, NULL, OSPopUp, EvButton, void);
    }
}

/*---------------------------------------------------------------------------*/

HWND _ospopup_focus_widget(OSPopUp *popup)
{
    cassert_no_null(popup);
    return popup->combo_hwnd;
}
